sklearn/linear_model/ridge.py

```diff
@@ -1015,48 +1164,168 @@ def _errors_and_values_helper(self, alpha, y, v, Q, QT_y):
             G_diag = G_diag[:, np.newaxis]
         return G_diag, c
 
-    def _errors(self, alpha, y, v, Q, QT_y):
-        G_diag, c = self._errors_and_values_helper(alpha, y, v, Q, QT_y)
+    def _errors_gram(self, alpha, y, v, Q, QT_y):
+        G_diag, c = self._errors_and_values_gram(alpha, y, v, Q, QT_y)
         return (c / G_diag) ** 2, c
 
-    def _values(self, alpha, y, v, Q, QT_y):
-        G_diag, c = self._errors_and_values_helper(alpha, y, v, Q, QT_y)
+    def _values_gram(self, alpha, y, v, Q, QT_y):
+        G_diag, c = self._errors_and_values_gram(alpha, y, v, Q, QT_y)
         return y - (c / G_diag), c
 
-    def _pre_compute_svd(self, X, y, centered_kernel=True):
-        if sparse.issparse(X):
-            raise TypeError("SVD not supported for sparse matrices")
-        if centered_kernel:
-            X = np.hstack((X, np.ones((X.shape[0], 1))))
+    def _decompose_covariance_sparse(self, X, y):
+        n_samples, n_features = X.shape
+        cov = np.empty((n_features + 1, n_features + 1))
+        cov[:-1, :-1], self._X_offset = self._compute_covariance(
+            X, self.fit_intercept)
+        if not self.fit_intercept:
+            cov = cov[:-1, :-1]
+        # to emulate centering X with sample weights,
+        # ie removing the weighted average, we add a column
+        # containing the square roots of the sample weights.
+        # by centering, it is orthogonal to the other columns
+        # when all samples have the same weight we add a column of 1
+        else:
+            cov[-1] = 0
+            cov[:, -1] = 0
+            if self._with_sw:
+                cov[-1, -1] = self._weight_sum
+            else:
+                cov[-1, -1] = n_samples
+        kernel_size = max(0, X.shape[1] - X.shape[0])
+        s, V = linalg.eigh(cov)
+        # remove eigenvalues and vectors in the null space of X^T.X
+        s = s[kernel_size:]
+        V = V[:, kernel_size:]
+        return s, V, X
+
+    def _errors_and_values_covariance_sparse_no_intercept(
+            self, alpha, y, s, V, X):
+        """compute loo values and dual coef when X is sparse"""
```

Comment:
Could you change slightly the docstring of the `_errors_and_values_...` functions such that they are all consistent. This is just a nitpick but like this we will have the same info in all of them.